courses/dss/percolator/src/server.rs

```diff
@@ -64,21 +73,57 @@ impl KvTable {
         ts_end_inclusive: Option<u64>,
     ) -> Option<(&Key, &Value)> {
         // Your code here.
-        unimplemented!()
+        let ts_start_inclusive = ts_start_inclusive.unwrap_or(0);
+        let ts_end_inclusive = ts_end_inclusive.unwrap_or(u64::MAX);
+        let map = match column {
+            Column::Write => &self.write,
+            Column::Data => &self.data,
+            Column::Lock => &self.lock,
+        };
+        map.range((
+            Included((key.clone(), ts_start_inclusive)),
+            Included((key, ts_end_inclusive)),
+        ))
+        .last()
+    }
+
+    #[inline]
+    fn contains_in_write_column(&self, primary: Vec<u8>, start_ts: u64) -> Option<u64> {
+        // Your code here.
+        for (key, value) in self.write.range((
+            Included((primary.clone(), 0)),
+            Included((primary, u64::MAX)),
+        )) {
+            match value {
+                Value::Timestamp(time) if *time == start_ts => return Some(*time),
```

Comment:
I think we should return commit timestamp in `key`, but not data timestamp in `value` here.

Reply:
Thanks for you careful review! You are right! I have fixed this bug.